Introduce first draft of Daikon++ with method ENTRY/EXIT invariant detection

erfan-arvan opened the pull request:
This PR introduces the first draft of **Daikon++** that detects and evaluates invariants at **method entry** and **method exit** program points. It implements an end-to-end pipeline:

- **Scan** Java sources for method ENTRY/EXIT points  
- **Generate** candidate invariants via LLM (scoped to parameters / return type) for each program point
- **Inject** one-line invariant guards into source files at the corresponding program point
- **Compile & Run** the target program to observe invariant outcomes  
  - Automatically **comment out** any injected invariants that cause compilation errors  
- **Parse logs** and report: held, falsified, failed-to-compile, and never-executed invariants

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Full CLI pipeline: LLM-backed invariant proposal → injection → compile/run → reporting; fat‑jar packaging; runtime invariant logging, deduplicating registry, quality filtering, and safe concurrent file writes.
  * Sample project and baseline-capture tooling for reproducible runs.

* **Documentation**
  * Expanded README with usage, configuration, and examples.

* **Tests**
  * New unit, integration, and regression tests exercising the end-to-end pipeline.

* **Chores**
  * Updated ignore rules (.idea/, .cache_daikonpp).
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

erfan-arvan committed:
Redesign Daikon++

erfan-arvan committed:
add README file

erfan-arvan committed:
update README file

erfan-arvan committed:
update README file

erfan-arvan committed:
update README file

erfan-arvan committed:
adding invariants in one line

erfan-arvan committed:
post compile filter

erfan-arvan committed:
fix a bug in collapseOnelineRegions

erfan-arvan committed:
fix a bug about literals

erfan-arvan committed:
Add support to distinguish compiled, executed, falsified, and held invariants

erfan-arvan committed:
fix a bug in injection

erfan-arvan committed:
fix a bug in injection

erfan-arvan committed:
fix a bug in injection

erfan-arvan committed:
fix a cfnullness error

erfan-arvan committed:
fix a minor error

erfan-arvan committed:
merge stderr into stdout for the run log file

erfan-arvan committed:
keep logs for FAILED-TO-COMPILE invariants by method

erfan-arvan committed:
Add end-to-end pipeline test and update Javadocs

kelloggm reviewed (commented):
I won't have time to go through this whole PR in detail until tomorrow, at the earliest. However, I can see already that I will complain about the lack of automated tests that mock the LLM. I'd like to at least see a version of the "full pipeline" test that mocks the LLM and which can be run on each change, but I'd prefer more than just the one smoke test.

erfan-arvan committed:
Add .coderabbit.yaml

erfan-arvan committed:
test(regression): add sample project baseline and automated replay test for Daikon++

erfan-arvan commented:
@kelloggm I have added the regression test:
I added a small sample Java project and ran the baseline capture script (baseline_capture.sh), which executes Daikon++ once and stores all generated artifacts (registry, logs, injected sources, working copy) under testdata/daikonpp_runs/baseline/.
I then added a JUnit test (RegressionReplayFromBaselineTest) that re-runs the full Daikon++ pipeline but skips the LLM calls, instead using the cached registry from the baseline to ensure reproducibility.
The test verifies that all newly generated files match the captured baseline exactly, hopefully ensuring that future code changes don’t alter deterministic parts of the pipeline.

erfan-arvan committed:
Fix formatting in .gitignore for build directory

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/JavaRunner.java:
```diff
@@ -0,0 +1,289 @@
+    Path errLog = classesDir.resolve("dp-javac.err");
+
+    // iterate: compile -> parse errors -> comment failing inv lines -> repeat
+    int pass = 1;
+    while (true) {
+      System.out.println(
+          ">>> Compiling with javac (pass " + pass + "): " + String.join(" ", baseCmd));
+      int code = runProcessNoThrow(baseCmd, srcRoot, outLog, errLog);
+
+      if (code == 0) {
+        System.out.println(">>> javac OK on pass " + pass);
+        return;
+      }
+
+      // parse errors and decide what to comment
+      String err = Files.exists(errLog) ? Files.readString(errLog) : "";
+      List<JavacError> errs =
+          parsePrimaryJavacErrors(err); // only lines like ".../Foo.java:123: error: ..."
+
+      // map -> comment
+      int commented = 0;
+      for (JavacError je : errs) {
+        Path file = Path.of(je.file).toAbsolutePath().normalize();
+        if (!file.startsWith(srcRoot)) continue; // ignore external/dep paths
+        commented += commentInvariantLine(file, je.line, je.message);
+      }
+
+      if (commented == 0) {
+        throw new RuntimeException(
+            "javac failed, but no one-line invariants were found to comment.\n" + err);
+      }
+
+      if (pass++ >= Math.max(1, maxPasses)) {
+        throw new RuntimeException("javac still failing after " + (pass - 1) + " passes.\n" + err);
+      }
```
`pass++` is a post-increment, so the old value of `pass` is used. This is a false positive.

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/results/.coderabbit.yaml:
```diff
@@ -0,0 +1,17 @@
+
+language_model:
+  provider: openai
+  model: gpt-4o-mini       # Or gpt-4o for more depth
```
I think it's right about the config not being read, since it's not obeying it.

kelloggm commented on src/test/java/edu/njit/jerse/daikonplusplus/RegressionReplayFromBaselineTest.java:
```diff
@@ -0,0 +1,272 @@
+      injector.injectGuards(e.getKey(), e.getValue());
+    }
+
+    // 5) Compile with javac and run with java to produce a fresh run log
+    Path classesDir = tmp.resolve("classes");
+    Files.createDirectories(classesDir);
+    String cp = System.getProperty("java.class.path");
+
+    JavaRunner.compileWithAutoFilter(workSrc, classesDir, cp, /*maxPasses*/ 10);
+
+    Path runLog = workSrc.resolve("daikonpp-run.log");
+    JavaRunner.run(MAIN_CLASS, JavaRunner.joinCp(cp, classesDir.toString()), List.of(), runLog);
```
Check for 11+

kelloggm commented on src/test/java/edu/njit/jerse/daikonplusplus/RegressionReplayFromBaselineTest.java:
```diff
@@ -0,0 +1,272 @@
+    // 6a) Compare injected sources
+    compareInjectedTrees(BASELINE_INJECTED, workSrc, "injected sources mismatch vs baseline");
+
+    // 6b) Compare run logs (exact)
+    assertTrue(Files.exists(BASELINE_RUNLOG), "Missing baseline run log: " + BASELINE_RUNLOG);
+    byte[] baseLog = Files.readAllBytes(BASELINE_RUNLOG);
+    byte[] newLog = Files.readAllBytes(runLog);
+    if (!Arrays.equals(baseLog, newLog)) {
+      // Helpful diff context
+      String baseS = new String(baseLog, StandardCharsets.UTF_8);
+      String newS = new String(newLog, StandardCharsets.UTF_8);
+      fail(
+          "Run log differs from baseline.\n--- baseline ---\n"
+              + snippet(baseS)
+              + "\n--- current ---\n"
+              + snippet(newS));
+    }
```
Always use `System.lineSeparator()` instead of `\n` for portability

kelloggm commented:
@erfan-arvan I've gone through the coderabbit comments and resolved the false positives. I've marked the ones that I think you should address with a 👍. I'll do a manual review of the tests and design next.

kelloggm commented on sample-project/src/sample/Main.java:
```diff
@@ -0,0 +1,16 @@
+package sample;
```
I would put the whole sample project in `src/test/resources` rather than in its own special location at top-level, to match Gradle's convention.

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/App.java:
```diff
@@ -1,11 +1,633 @@
+ * scan → LLM invariants → inject → <b>javac</b> → <b>java</b> → parse log → print held ENTRY
+ * invariants per method.
+ */
+public final class App {
```
Give this a better name. Maybe `DaikonPlusPlus`?

kelloggm replied:
+1

kelloggm commented on src/test/java/edu/njit/jerse/daikonplusplus/AppTest.java:
```diff
@@ -1,13 +1,94 @@
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+final class AppTest {
```
Give this a better name. It appears to be a small collection of unit tests?

kelloggm commented on src/test/java/edu/njit/jerse/daikonplusplus/RegressionReplayFromBaselineTest.java:
```diff
@@ -0,0 +1,272 @@
+import org.junit.jupiter.api.io.TempDir;
+
+/**
+ * Regression test that replays the baseline registry.jsonl (stored in the repo) and then runs the
```
This isn't really a regression test so much as it is a smoke test: this test is a single, high-level check that the pipeline works on the "golden path" (i.e., the conditions under which we expect the program to run). This is a pretty weak control on the program's logic; I suspect that there are many things that we could break in this implementation that this test won't catch.

Especially for LLM-generated code (which I know a lot of this to be), it's usually a good idea to have a strong test suite. What I'd suggest is to build an _infrastructure_ (of which this test is but one instantiation) to run the full pipeline on target programs. Then, as you fix the bugs you mentioned to me yesterday that you've encountered when running Daikon++ on the CF, you can add the triggering code as a new sample project with an associated test case. In other words, design this so that it's easy to make new test cases with only a little effort.

You might want to look at how Specimin's test infrastructure works: https://github.com/njit-jerse/specimin/blob/main/src/test/java/org/checkerframework/specimin/SpeciminTestExecutor.java

kelloggm replied:
(Note how the other files in the same directory - the specific Specimin tests - are subclasses of the test executor. I suggest a similar design for this project.)

kelloggm commented on testdata/daikonpp_runs/baseline/app-console.log:
```diff
@@ -0,0 +1,36 @@
+>>> Working copy created at: /Users/erfanarvan/daikonplusplus/build/daikonpp_work/src-20251024-133426
```
These logs should get sanitized before they're committed. Also, it's critical not to require that these paths exactly match: the tests should run the same whether they're running on my machine or yours.

kelloggm commented on src/test/java/edu/njit/jerse/daikonplusplus/PipelineWithRealLlmIT.java:
```diff
@@ -0,0 +1,157 @@
+import org.junit.jupiter.api.Timeout;
+import org.junit.jupiter.api.io.TempDir;
+
+/**
```
Make sure that this test is disabled by default when you run `./gradlew test`. Then, set up Github Actions CI that runs `./gradlew test` to make sure that the smoke and regression tests are always passing.

erfan-arvan committed:
add end-to-end test infrastructure with LLM cassette replay support

erfan-arvan commented:
@kelloggm 

I added an end-to-end testing infrastructure with deterministic replay using stored LLM cassettes (each cassette is a JSON file that records the exact input–output pair from a real LLM call, named by a stable hash of the prompt for consistent lookup and replay). To add a new test, we create a folder under `src/test/resources/daikonpp-pipeline/<case>/` that contains an input/ directory (Java sources) and config.json (which specifies the main class and optional parameters like maxK). Then we run the `scripts/record_one.sh <case>` script, which executes the pipeline with the real LLM, generates the expected/ snapshots, and stores the corresponding LLM cassettes under `src/test/cassettes/`.

During actual tests, the pipeline replays responses from these cassettes and compares the new outputs with the expected files for each test case. I have added two sample tests for now and will include more real ones as I address your comments, address the problems reported by the CodeRabbit you marked that should be addressed, and continue running Daikon++ on real projects to fix bugs

kelloggm commented on build.gradle:
```diff
@@ -42,16 +42,51 @@ dependencies {
     checkerFramework 'org.checkerframework:checker:3.49.5'
     compileOnly      'org.checkerframework:checker-qual:3.49.5'
 
+    // --- TEST ONLY ---
```
This comment is unnecessary, because the code itself already carries the information that mockito is a test-only dependency in the name of the configuration (`testImplementation`). In general, avoid comments that are duplicative with the code; the purpose of comments is to expand on the code to give a human reader more context (e.g., _why_ is the code written this way, what alternatives were considered, etc.).

LLMs especially love to write these kind of inane comments. If you're using one for code generation, configure it to avoid comments like these (or delete them yourself). Comments like this one are just clutter.

kelloggm commented on scripts/new-baseline.sh:
```diff
@@ -0,0 +1,105 @@
+# Usage:
+#   scripts/new-baseline.sh /path/to/src_root [case-name] [maxK]
+#
+# Examples:
```
These examples aren't very useful without a description of what this script is supposed to do. Add a comment explaining its purpose.

kelloggm commented on scripts/new-baseline.sh:
```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env bash
```
If we're going to include bash scripts, add a [shellcheck](https://www.shellcheck.net/) job to the gradle build. It's a useful linter, and IMO effectively required for any project that involves shell scripting.

kelloggm commented on scripts/record_one.sh:
```diff
@@ -0,0 +1,117 @@
+#
+# Environment variables:
+#   DP_LLM_CASSETTES   Directory for storing LLM interaction logs (set automatically)
+#   DP_DISABLE_REAL_LLM If set to 1, disables real LLM calls (should be UNSET for recording)
```
This doesn't match the script's actual behavior, which always unsets this variable.

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/config/DpConfig.java:
```diff
@@ -68,34 +70,50 @@ public boolean keepWork() {
-    if (v == null) return def;
+  // ---- helpers ----
 
+  private static boolean getBool2(
```
Why `getBool2` and not just `getBool`? You've deleted the old version.

kelloggm reviewed (commented):
This review is just of the cassette infrastructure. Overall, it looks good to me.

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/App.java:
```diff
@@ -1,11 +1,660 @@
+
+  private static final boolean DEBUG = CFG.debug();
+
+  private static final java.util.Set<String> RUN_DEDUP =
```
Why use the fully-qualified name of `Set` here?

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/App.java:
```diff
@@ -1,11 +1,660 @@
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.*;
```
Configure the automated formatter to avoid wildcard imports (they create confusion about what classes, exactly, are being imported, so good style guides forbid them).

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/App.java:
```diff
@@ -1,11 +1,660 @@
+  private static final java.util.Set<String> RUN_DEDUP =
+      java.util.concurrent.ConcurrentHashMap.newKeySet();
+
+  private static String keyFor(ProgramPoint pt, String expr) {
```
nit: I usually prefer to put utility methods like this one in their own class, or in the class on which they operate, rather than in the class where they are used. For this one, I'd make it `public` (I see no reason to hide it) and put it in `ProgramPoint`

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/App.java:
```diff
@@ -1,11 +1,660 @@
+
+    final List<String> programArgs = new ArrayList<>();
+    if (i < args.length && args[i].equals("--")) {
+      for (i = i + 1; i < args.length; i++) programArgs.add(args[i]);
```
nit: avoid brace-less control flow constructs, to avoid the possibility of bugs like [this one](https://www.blackduck.com/blog/understanding-apple-goto-fail-vulnerability-2.html)

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/App.java:
```diff
@@ -1,11 +1,660 @@
-  public static void main(String[] args) {
-    System.out.println(new App().getGreeting());
+  public static void main(String[] args) throws Exception {
+    if (args.length < 3) {
```
Argument parsing is surprisingly easy to mess up, and often it's common to want to add more arguments to a program later. For that reason, I usually use a library for options parsing. Since you've already implemented it here, I wouldn't replace it _right now_, but you should add a TODO that says that we should replace this manual argument parsing logic with a library the next time we want to add an option to DaikonPlusPlus.

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/App.java:
```diff
@@ -1,11 +1,660 @@
+    final DpConfig cfg = DpConfig.fromEnv();
+    if (CFG.registryReset()) {
+      try {
+        java.nio.file.Files.deleteIfExists(cfg.registryPath());
```
Probably a good idea to back this up, by default, rather than just deleting it entirely.

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/App.java:
```diff
@@ -1,11 +1,660 @@
+    // --- Phase 3: compile with javac and run with java
+    final Path classesDir = srcRoot.resolve("daikonpp-classes"); // output dir for compiled classes
+    final String selfCp =
+        System.getProperty("java.class.path"); // include our runtime (InvariantLogger)
```
If we're going to include our own run-time dependencies in the classpath that we use to run the target, we probably want to shadow ourselves first (i.e., move all of our classes and our dependencies' classes into new packages, to avoid conflicts). This isn't urgent until we try to run on a program that shares a dependency with us but needs a conflicting version, so just add a TODO here for now acknowledging that we're aware of the problem and that we will eventually need to fix it.

There's a discussion of how to do this here that might be a good starting point to link in the TODO: https://softwareengineering.stackexchange.com/questions/297276/what-is-a-shaded-java-dependency

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/App.java:
```diff
@@ -1,11 +1,660 @@
+      } else {
+        verdict =
+            InvariantRegistry.Verdict
+                .PROPOSED; // seen in registry but neither compiled nor executed
```
Should this case be possible? If not, it may be better to throw an exception.

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/App.java:
```diff
@@ -1,11 +1,660 @@
+    System.out.println(">>> Outcomes: " + cfg.outcomesPath().toAbsolutePath());
+
+    // observed-held = executed − falsified   (only those we observed to run and not fail)
+    Map<String, List<edu.njit.jerse.daikonplusplus.App.RecordLite>> heldByMethod = new TreeMap<>();
```
FQNs are bad enough for imported library classes, but for classes _in this file_ it seems a bit extreme.

kelloggm commented on src/main/java/edu/njit/jerse/daikonplusplus/App.java:
```diff
@@ -1,11 +1,660 @@
+    // observed-held = executed − falsified   (only those we observed to run and not fail)
+    Map<String, List<edu.njit.jerse.daikonplusplus.App.RecordLite>> heldByMethod = new TreeMap<>();
+    // falsified ∩ all (to ignore any stray ids not in registry)
+    Map<String, List<edu.njit.jerse.daikonplusplus.App.RecordLite>> falsByMethod = new TreeMap<>();
```
Why are these `TreeMap`s? That seems like an oddly-specific choice of data structure; I would've just used a `HashMap`.

kelloggm reviewed (changes requested):
Review of `App#main` only.
